### ISSUE TITLE
LLM Observability: Add optional Langfuse support

### ISSUE THREAD

**Issue by aiakubovich** [Azure-Samples/azure-search-openai-demo#1300]
Hi! We use the app extensively, but it lacks LLM observability functionality. There should be an easy way to gather user interactions within the AOAI app. [Langfuse](https://github.com/langfuse/langfuse) is an open-source LLM observability platform that supports prompt logging. It offers [native integration with OpenAI library](https://langfuse.com/docs/openai) which can be achieved by adding just a few lines of code to the underlying Python script. I am wondering whether it would be feasible to add optional Langfuse support into the app, as it seems that only minimal changes are required (a few variables for the Azure Web App and few lines in the AOAI app code). If this is ok, I can attempt to create a pull request for it.

**Comment by pamelafox** [comment]
Very interesting, I hadn't seen that before, and there's a lot of interest in observability. I think we'd want to get Langfuse deployed on Azure to make sure it was viable for devs to host it on our infrastructure. It looks like that'd be fairly straightforward using Azure Container Apps and the Postgres ACA Add-on, for example. Are you hosting Langfuse on Azure or using their cloud-based version?

**Comment by aiakubovich** [comment]
@pamelafox, we are hosting Langfuse on Azure as Azure Web App and using Azure PostgreSQL - Flexible Server. Azure Container Apps would be another option. It was very simple to deploy, here is the [guide](https://langfuse.com/docs/deployment/self-host).

**Comment by tonybaloney** [comment]
Have a look at https://github.com/Azure-Samples/azure-search-openai-demo/issues/1318 as well, we're considering adding this to improve the Application Insights traces for OpenAI

**Comment by marcklingen** [comment]
Langfuse maintainer here. There are many teams running Langfuse on Azure in prod, mostly on (1) Azure Container Instances + Azure PostgreSQL or (2) everything in Kubernetes.

Let me know if I can help with anything here @aiakubovich @pamelafox

**Comment by marcklingen** [comment]
> From my experiment, the changes to this repo are pretty minimal - just setting the env variables and changing the import in app.py - and we can just document those in docs/.

Let me know if there's something I can help with.

> What I'm more interested in is making it super easy to deploy langfuse to Azure using the Azure Developer CLI (`azd up`). I've started that at https://github.com/pamelafox/langfuse-azure but have some TODOs. One thing that would help is if we could specify the database password separately from the URL, so that we can store those with secret-specific mechanisms. I could also store the entire database URI as a secret, but it's nice when you can view the rest in plaintext.

That's awesome, I'd love to reference https://github.com/pamelafox/langfuse-azure in our self-hosting docs to make it easier for others to get started on Azure (once you are done with it) – if that's ok for you.

> Instead of DATABASE_URL, you can also use DATABASE_HOST, DATABASE_USERNAME, DATABASE_PASSWORD and DATABASE_NAME.
> Self hosting docs

This was added to langfuse specifically to support external secret management to e.g. rotate secrets easily on Kubernetes. Does this solve for your use case on Azure to specify `DATABASE_PASSWORD` independently?

**Comment by pamelafox** [comment]
Thank you! I've got the basic deployment working now. Next step is to add in authentication via AAD, as Azure customers will prefer that, I think. 

@aiakubovich Did you all go with AAD for your deployment? Just curious.

**Comment by aiakubovich** [comment]
@pamelafox, yes, we added few env variables from Azure Active Directory to the Azure Web App config page based on Langfuse self-host guide and now we able to see Azure button on Langfuse UI and it works well for authentication via AD.  

Also question regarding this repo - any plans in changing the import in app.py to support Langfuse by default?

**Comment by pamelafox** [comment]
Here are some screenshots from the langfuse on ACA deployment:

![Screenshot 2024-02-29 at 2 41 06 PM](https://github.com/Azure-Samples/azure-search-openai-demo/assets/297042/0c394e3e-d8ca-40e7-8c64-d103876c4b0a)
![Screenshot 2024-02-29 at 2 40 54 PM](https://github.com/Azure-Samples/azure-search-openai-demo/assets/297042/182123a5-d4b7-4c59-a1bf-867a9761008d)
![Screenshot 2024-02-29 at 2 40 46 PM](https://github.com/Azure-Samples/azure-search-openai-demo/assets/297042/6500a815-7fa5-4ef7-b853-ed728d09da31)

That's from this repo:
https://github.com/pamelafox/langfuse-azure
Still WIP as you can see from the TODOs, but basic deployment works.

Question for @marcklingen: Is it expected that the traces show null for input/output/metadata?

@aiakubovich We wouldn't make that change across the board, as we want all add-ons like this to be optional (even Azure monitor is optional, though we now default to true as its negligible costs). We could potentially add it as an option based off an env variable, which would look something like this code inside setup_clients:

```
if (os.getenv("LANGFUSE_HOST")):
   from langfuse.openai import AsyncOpenAI, AsyncAzureOpenAI
```

For now I'd like to get langfuse-azure template done, document that, and see what feedback we get.

**Comment by pamelafox** [comment]
@marcklingen Also wondering why "Time to first token" would be empty?

**Comment by pamelafox** [comment]
FYI, I showed Langfuse on Azure in a livestream today-
https://youtube.com/live/Q169GFD_9PI?feature=share
The repo is here:
https://github.com/Azure-Samples/langfuse-on-azure

I'll add documentation here about that repo shortly.

**Comment by marcklingen** [comment]
@pamelafox Just checked out the livestream, thanks for featuring and integrating https://github.com/langfuse!

> Question for @marcklingen: Is it expected that the traces show null for input/output/metadata?

Yes, the openai monkeypatch integration is the easiest to get started with to log single generations (llm calls, embeddings) to Langfuse. Most users expand to complete traces with more spans (eg for retrieval) or additional events (eg other non-llm api calls). See a full guide here: https://langfuse.com/docs/tracing

> Also wondering why "Time to first token" would be empty?

Good one, just created an issue to add this to our openai integration when streaming: https://github.com/langfuse/langfuse/issues/1330

**Comment by marcklingen** [comment]
By the way, just added a link to [Azure-Samples/langfuse-on-azure](https://github.com/Azure-Samples/langfuse-on-azure) to our self-hosting guide to make deploying to Azure easier for others as well. thanks again!

https://langfuse.com/docs/deployment/self-host#azure
Change: https://github.com/langfuse/langfuse-docs/commit/d81a69e8d647f4fef0f607eba0c53b0756f783a9

**Comment by bsonnek** [comment]
Hi @marcklingen and @pamelafox - This looks like an older thread, but I was wondering if anyone has been able to get the new Langfuse Observe decorators working with the llm 'generation' in this backend app of this repo?

 I'm new to decorators and the methods used to call the Azure OpenAI chat and embeddings in this enterprise chat app. I'm struggling to get the observe decorator to show token usage in this backend app. 
Thanks for all that you both do!

**Comment by bsonnek** [comment]
Hi @marcklingen 
I ended up finding a very simple solution after reading through the Langfuse documentation: https://langfuse.com/docs/integrations/openai/python/structured-outputs

All I had to do is change this line:
https://github.com/Azure-Samples/azure-search-openai-demo/blob/009d5e13ba5f9f14bcca684d7a47e1d28fd9023e/app/backend/app.py#L31

From: 
from openai import AsyncAzureOpenAI, AsyncOpenAI
To:
from langfuse.openai import AsyncAzureOpenAI, AsyncOpenAI

Now I am seeing the generations in the langfuse dashboard.
Thanks again for your quick response.
